ports/nrf/mpconfigdevice_nrf51822.h

```diff
@@ -0,0 +1,49 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 Glenn Ruben Bakke
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+// Board overridable build configuration.
+
+#ifndef MICROPY_MBFS
+#define MICROPY_MBFS                       (1)
+#endif
+
+#ifndef MICROPY_VFS
+#define MICROPY_VFS                        (0)
+#endif
+
+// Board overridable feature configuration.
+
+#ifndef MICROPY_PY_ARRAY_SLICE_ASSIGN
+#define MICROPY_PY_ARRAY_SLICE_ASSIGN      (!defined(BLUETOOTH_SD))
```

Comment:
it's not allowed to use `defined` within a `#define`; instead do:
```c
#define MICROPY_PY_ARRAY_SLICE_ASSIGN      (!BLUETOOTH_SD)
```
or, if the symbol`BLUETOOTH_SD` is not guaranteed to be defined, probably it should be:
```c
#if defined(BLUETOOTH_SD)
#define MICROPY_PY_ARRAY_SLICE_ASSIGN      (0)
#else 
#define MICROPY_PY_ARRAY_SLICE_ASSIGN      (1)
#endif 
```

Reply:
> it's not allowed to use defined within a #define

I did some tests here, and apparently it looks like it does work. Could it be because `BLUETOOTH_SD` is passed through the compiler as a `CFLAG`? 

However, i get a few 12 extra bytes in the compilation when using `(!defined(BLUETOOTH_SD))` compared to just set 0 or 1. I cross checked also thatLTO and DEBUG has no impact on this. It would have looked cleaner doing a single line. I guess, until i can explain those 12 extra bytes, it is more safe to do what you suggest here, doing the second variant as BLUETOOTH_SD might not be present.

I'll post the results here in case it might be worth something (I also cross checked with pca10056, and it is similar compilations).

**make BOARD=microbit** 
```
mpconfigdevice_nrf51822.h:
    #define MICROPY_PY_UBINASCII (!defined(BLUETOOTH_SD))

LINK build-microbit/firmware.elf
   text	   data	    bss	    dec	    hex	filename
 142220	    100	   1156	 143476	  23074	build-microbit/firmware.elf

firmware.map: 
                0x000000000001d67c                mp_module_ubinascii
```
```
mpconfigdevice_nrf51822.h:
    #define MICROPY_PY_UBINASCII (1)

LINK build-microbit/firmware.elf
   text	   data	    bss	    dec	    hex	filename
 142232	    100	   1156	 143488	  23080	build-microbit/firmware.elf

firmware.map:
                0x000000000001d67c                mp_module_ubinascii
```
```
mpconfigdevice_nrf51822.h:
    #define MICROPY_PY_UBINASCII (0)

LINK build-microbit/firmware.elf
   text	   data	    bss	    dec	    hex	filename
 141264	    100	   1156	 142520	  22cb8	build-microbit/firmware.elf

firmware.map: Symbol mp_module_ubinascii not found
```




**make BOARD=microbit SD=s110** 
```
mpconfigdevice_nrf51822.h:
    #define MICROPY_PY_UBINASCII (!defined(BLUETOOTH_SD))

LINK build-microbit-s110/firmware.elf
   text	   data	    bss	    dec	    hex	filename
 149088	    100	   1300	 150488	  24bd8	build-microbit-s110/firmware.elf

firmware.map: Symbol mp_module_ubinascii not found
```
```
#define MICROPY_PY_UBINASCII (1)

LINK build-microbit-s110/firmware.elf
   text	   data	    bss	    dec	    hex	filename
 150048	    100	   1300	 151448	  24f98	build-microbit-s110/firmware.elf

firmware.map: 
                0x0000000000036a00                mp_module_ubinascii
```
```
mpconfigdevice_nrf51822.h:
    #define MICROPY_PY_UBINASCII (0)

LINK build-microbit-s110/firmware.elf
   text	   data	    bss	    dec	    hex	filename
 149100	    100	   1300	 150500	  24be4	build-microbit-s110/firmware.elf

firmware.map: Symbol mp_module_ubinascii not found
```

Reply:
> I did some tests here, and apparently it looks like it does work

It could just be gcc that allows it.  I'm pretty sure it's invalid code by the C standard and clang or other compilers will reject it.  Might not be an issue for the nrf port with `arm-none-eabi-gcc` but IMO better to avoid this pattern.



> However, i get a few 12 extra bytes in the compilation when using `(!defined(BLUETOOTH_SD))` compared to just set 0 or 1

From the output above it looks like the 12 extra bytes are there when using 0/1 (ie the opposite of your statement).... and that could be just because the git repo is dirty and it adds the `-dirty` string to the REPL banner, which increases code size.

Reply:
> From the output above it looks like the 12 extra bytes are there when using 0/1 (ie the opposite of your statement).... and that could be just because the git repo is dirty and it adds the -dirty string to the REPL banner, which increases code size.

Yeah, that makes sense! Thanks! I had forgot about the appended string. The count also indicates this.